src/Compilers/Core/Portable/Syntax/SyntaxNavigator.cs

```diff
@@ -484,34 +473,24 @@ internal SyntaxToken GetLastToken(SyntaxNode current, Func<SyntaxToken, bool> pr
         {
             while (node.Parent != null)
             {
-                // walk forward in parent's child list until we find ourselves and then return the
-                // previous token
-                bool returnPrevious = false;
-                foreach (var child in node.Parent.ChildNodesAndTokens().Reverse())
+                foreach (var child in node.SiblingsBeforeSelf())
```

Comment:
i actuqally like the SiglingsBefore/AfterSelf bit.  Can you extract that into a dedicated PR that will be easy to review.  We can then look at the rest in isolation.